addons/mass_mailing/models/mailing.py

```diff
@@ -1117,15 +1117,21 @@ def _action_send_mail(self, res_ids=None):
             ).create(composer_values)
 
             # auto-commit except in testing mode
-            composer._action_send_mail(
-                auto_commit=not getattr(threading.current_thread(), 'testing', False)
-            )
+            current_thread = threading.current_thread()
+            auto_commit = not (hasattr(current_thread, 'testing') and current_thread.testing)
+            composer._action_send_mail(auto_commit=auto_commit)
+
             mailing.write({
                 'state': 'done',
                 'sent_date': fields.Datetime.now(),
                 # send the KPI mail only if it's the first sending
                 'kpi_mail_required': not mailing.sent_date,
             })
+
+            # ensure mailing state update after auto-commit
+            if auto_commit is True:
+                self._cr.commit()
```

Comment:
Model._cr, Model._context, etc.. are being deprecated
```suggestion
                self.env.cr.commit()
```